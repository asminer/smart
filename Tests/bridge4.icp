


/*
	Bridge crossing optimization problem, it goes something like this.
	There are 4 people that must cross an old bridge at night,
	and only 1 flashlight.  2 people can cross the bridge at a time.
	Each person walks at a different speed.  What is the fastest
	we can get everyone across the bridge?

	People start on the left side and go to the right side.

	We identify people by their speeds, so the speeds
	need to be unique.
*/

int L1a in {1};		// speed of person 1
int L1b in {2};		// speed of person 2
int L1c in {5}; 	// speed of person 3
int L1d in {10};	// speed of person 4

int L2a, L2b, R2a, R2b	in {1, 2, 5, 10};	// after 1 l-r cross
int L3a, L3b, L3c, R3a	in {1, 2, 5, 10};	// l-r, r-l
int L4a, R4a, R4b, R4c	in {1, 2, 5, 10};	// l-r, r-l, l-r
int L5a, L5b, R5a, R5b	in {1, 2, 5, 10};	// right before last cross
int R6a, R6b, R6c, R6d	in {1, 2, 5, 10};	// everyone across

// First, we make sure that everyone is accounted for at each state.

// State 1 unique
L1a != L1b;
L1a != L1c;
L1a != L1d;
L1b != L1c;
L1b != L1d;
L1c != L1d;

// State 2 unique
L2a != L2b;
L2a != R2a;
L2a != R2b;
L2b != R2a;
L2b != R2b;
R2a != R2b;

// State 3 unique
L3a != L3b;
L3a != L3c;
L3a != R3a;
L3b != L3c;
L3b != R3a;
L3c != R3a;

// State 4 unique
L4a != R4a;
L4a != R4b;
L4a != R4c;
R4a != R4b;
R4a != R4c;
R4b != R4c;

// State 5 unique
L5a != L5b;
L5a != R5a;
L5a != R5b;
L5b != R5a;
L5b != R5b;
R5a != R5b;

// State 6 unique
R6a != R6b;
R6a != R6c;
R6a != R6d;
R6b != R6c;
R6b != R6d;
R6c != R6d;

// First crossing
(R2a == L1a) | (R2a == L1b) | (R2a == L1c) | (R2a == L1d);
(R2b == L1a) | (R2b == L1b) | (R2b == L1c) | (R2b == L1d);
(L2a == L1a) | (L2a == L1b) | (L2a == L1c) | (L2a == L1d);
(L2b == L1a) | (L2b == L1b) | (L2b == L1c) | (L2b == L1d);

// Second crossing
L3a == L2a;
L3b == L2b;
(L3c == R2a) | (L3c == R2b);
(R3a == R2a) | (R3a == R2b);

// Third crossing
(L4a == L3a) | (L4a == L3b) | (L4a == L3c);
R4a == R3a;
(R4b == L3a) | (R4b == L3b) | (R4b == L3c);
(R4c == L3a) | (R4c == L3b) | (R4c == L3c);

// Fourth crossing
L5a == L4a;
(L5b == R4a) | (L5b == R4b) | (L5b == R4c);
(R5a == R4a) | (R5a == R4b) | (R5a == R4c);
(R5b == R4a) | (R5b == R4b) | (R5b == R4c);

// Fifth crossing
R6a == R5a;
R6b == R5b;
(R6c == L5a) | (R6c == L5b);
(R6d == L5a) | (R6d == L5b);

// These are used to eliminate symmetries, 
// and to make the "reward" function easier to write.

R2a < R2b;	// the two people who crossed;
R4b < R4c;	// WLOG we can order them.
R6c < R6d;

minimize cost := R2b + L3c + R4c + L5b + R6d;

