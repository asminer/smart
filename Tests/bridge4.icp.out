Minimum value 17 obtainted in state [L1a = 1, L1b = 2, L1c = 5, L1d = 10, L2a = 5, L2b = 10, R2a = 1, R2b = 2, L3a = 5, L3b = 10, L3c = 1, R3a = 2, L4a = 1, R4a = 2, R4b = 5, R4c = 10, L5a = 1, L5b = 2, R5a = 5, R5b = 10, R6a = 5, R6b = 10, R6c = 1, R6d = 2]
cost: 17
